internal/action/verify.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package action
+
+import (
+	"errors"
+
+	"github.com/opencontainers/go-digest"
+	helmaction "helm.sh/helm/v3/pkg/action"
+	helmchart "helm.sh/helm/v3/pkg/chart"
+	helmchartutil "helm.sh/helm/v3/pkg/chartutil"
+	helmrelease "helm.sh/helm/v3/pkg/release"
+
+	helmdriver "helm.sh/helm/v3/pkg/storage/driver"
+
+	helmv2 "github.com/fluxcd/helm-controller/api/v2beta2"
+	"github.com/fluxcd/helm-controller/internal/chartutil"
+	"github.com/fluxcd/helm-controller/internal/release"
+)
+
+var (
+	ErrReleaseDisappeared = errors.New("observed release disappeared from storage")
+	ErrReleaseNotFound    = errors.New("no release found")
+	ErrReleaseNotObserved = errors.New("release not observed to be made by reconciler")
+	ErrReleaseDigest      = errors.New("release digest verification error")
+	ErrChartChanged       = errors.New("release chart changed")
+	ErrConfigDigest       = errors.New("release config digest verification error")
+)
+
+// VerifyStorage verifies that the last release in the Helm storage matches the
+// Current state of the given HelmRelease. It returns the release, or an error
+// of type ErrReleaseDisappeared, ErrReleaseNotFound, ErrReleaseNotObserved, or
+// ErrReleaseDigest.
+func VerifyStorage(config *helmaction.Configuration, obj *helmv2.HelmRelease) (*helmrelease.Release, error) {
+	curRel := obj.Status.Current
+	rls, err := config.Releases.Last(obj.GetReleaseName())
+	if err != nil {
+		if errors.Is(err, helmdriver.ErrReleaseNotFound) {
+			if curRel != nil && curRel.Name == obj.GetReleaseName() && curRel.Namespace == obj.GetReleaseNamespace() {
+				return nil, ErrReleaseDisappeared
+			}
+			return nil, ErrReleaseNotFound
+		}
+		return nil, err
+	}
+	if curRel == nil {
```

Comment:
shouldn't we check this before checking if a release has disappeared?

Reply:
No, because for a release not to be observed it has to exist. So we do everything that depends on the error (which also confirms existence) first.